DEWP: Handle cyclical module dependencies

michalczaplinski opened the pull request:
## What?
Fixes: https://github.com/WordPress/gutenberg/issues/65288

## How?
By tracking which dependencies have already been visited.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

michalczaplinski committed:
Add tests for cyclical dependencies

sirreal commented on packages/dependency-extraction-webpack-plugin/test/fixtures/cyclic-deps/webpack.config.js:
```diff
@@ -0,0 +1,8 @@
+const DependencyExtractionWebpackPlugin = require( '../../..' );
+
+module.exports = {
+	plugins: [ new DependencyExtractionWebpackPlugin() ],
```
I think there will need to be an external module listed here that's used in one or both of the file and we can probably disable defaults as well.

michalczaplinski replied:
Thanks Jon! Yeah, I've added it now! 

I think I didn't realize this before: The problem that I found only affects the **externalized** modules. Do you think that's correct?

michalczaplinski committed:
Add the externals and make the test fail

michalczaplinski commented:
I was able to reproduce the error (originally found in [here](https://github.com/WordPress/gutenberg/actions/runs/10831421519/job/30053367698?pr=64067#step:4:84)) in a [new test case](https://github.com/WordPress/gutenberg/actions/runs/10883556999/job/30196982817?pr=65291#step:6:338)

michalczaplinski committed:
Add cycle detection in dependency path checking

Enhanced the dependency resolution logic to detect cycles in the module graph, preventing infinite loops during static dependency path checks. Introduced a Set to track visited blocks and avoid revisiting them.

michalczaplinski commented on packages/dependency-extraction-webpack-plugin/lib/index.js:
```diff
@@ -352,12 +352,25 @@ class DependencyExtractionWebpackPlugin {
-	 * @param {webpack.DependenciesBlock} block
+	 * @param {webpack.Compilation}            compilation
+	 * @param {webpack.DependenciesBlock}      block
+	 * @param {Set<webpack.DependenciesBlock>} visited
```
It might be better to only store IDs of each visited block instead of the whole block. This would use less memory.  

However, I could not see if the `DependenciesBlock` has a stable ID of any kind: https://github.com/webpack/webpack/blob/899f06934391baede59da3dcd35b5ef51c675dbe/lib/DependenciesBlock.js#L29

Without benchmarking, it's also hard to know if it makes a difference.

sirreal replied:
I have some ideas for this, I'm working on it now.

sirreal committed:
Revert changes

sirreal committed:
Propose static WeakSet/WeakMap implementation.

sirreal committed:
Add CHANGELOG entry

sirreal committed:
Remove redundant plugin config in test

sirreal committed:
Revert "Remove redundant plugin config in test"

This reverts commit b5e33dbc0e6db4d4aab07de4d52b42ae6b9672ee.

sirreal committed:
Remove redundant plugin config in test

sirreal commented on packages/dependency-extraction-webpack-plugin/lib/index.js:
```diff
@@ -349,6 +349,9 @@ class DependencyExtractionWebpackPlugin {
 		}
 	}
 
+	static #staticDepsCurrent = new WeakSet();
+	static #staticDepsCache = new WeakMap();
```
@michalczaplinski I used the same basic idea but with a few tweaks:

- WeakSet for what's currently being explored. If the block is being explored and is being searched again, it does not represent a static path to the root. This should allow traversal to continue along other paths even if ultimately the current block is a static path to the root.
- WeakMap for what's already been explored. This shouldn't be strictly necessary, but serves as a cache to avoid traversing the same paths over and over.

These weak data structures should work very well here, as long as the webpack compilation holds a reference to the blocks they can be cached, but will be garbage collected when webpack no longer holds a ref.

The current set could likely be a regular set since we should always clean up, but I don't think there's harm in it being weak.

michalczaplinski replied:
Nice! Yeah, the weak structures work well here 👍 

The cache is a nice touch. We have a space-time complexity tradeoff, and I agree that using more memory in favor of time makes sense.

michalczaplinski committed:
Updated the snapshot files

michalczaplinski commented:
@sirreal Had to update the test snapshots in https://github.com/WordPress/gutenberg/pull/65291/commits/81d15af7118b2e5ec7b5c556d4e2cbc3b3ca11f7 for unit tests to pass.

Could you double-check that the externals type "module" is what we want instead of "import" (according to the test snapshot)?

sirreal commented:
Yep, looks right 👍

michalczaplinski committed:
Merge remote-tracking branch 'origin/trunk' into fix/dewp-cyclical-deps
